### DISCUSSION TITLE
Can you send and receive websocket messages at the same time from different threads?

### DISCUSSION THREAD

**Discussion by jamespinkerton** [curl/curl#15484] [General]
Hi. If I open a websocket connection with a single handler using the curl websocket API and I manage the event loop myself, I can call send and recv functions. In theory, send is writing to the network card write buffer, and recv is reading from the read buffer. So I probably should be able to both send and receive messages at the same time. However I know curl advises never to use the same handler from two threads. Does anyone know what the practical realities are about doing this?

Thanks so much
James

**Comment by SCRIER-org** [comment]
Briefly, you dedicate one thread to handling I/O packets.  And then you feed its outgoing stream with an outbound threadsafe queue.  Any thread that wants to send a one-way message, packages it up into your custom Message struct, then pushes it on the back of the queue.  And your I/O handler loop keeps popping Messages off the front of the queue, and using their payload and length contents to shove them out the door.  In this way you support _sending_ of one-way outbound messages from multiple threads, using a single-threaded handler.

_Receiving_ is up for grabs.  You in theory could do dispatching from inside your handler thread.  But it is probably conceptually easier to set up a second thread-safe queue, and then package and shove received messages into it as fast as they come in.  Then you can have a dedicated Dispatcher thread on the other end of the queue, with a "kick me in the butt and wake me up when I'm sleeping" cv flag, that pulls the messages off the incoming queue stuffed by your handler, and hands them out like Christmas presents to deserving threads.  This handles one-way (streaming) message _receiving_.

Request/sometime-later-Reply two-way messages require a little bit more smarts in the Dispatcher (better tags, "To Zaphod"), and usually future/promise --blocking /OR non-blocking (tasting)-- receiving assignments on the part of your various little Application threads.

Good luck.
